logstash-core/src/main/java/org/logstash/Event.java

```diff
@@ -213,6 +240,62 @@ public void setField(final FieldReference field, final Object value) {
         }
     }
 
+    private boolean isTagsWithMap(final FieldReference field) {
+        return field.getPath() != null && field.getPath().length > 0 && field.getPath()[0].equals(TAGS);
+    }
+
+    private boolean isLegalTagValue(final Object value) {
+        if (value instanceof String || value instanceof RubyString || value == null) {
+            return true;
+        } else if (value instanceof List) {
+            for (Object item: (List) value) {
+                if (!(item instanceof String) && !(item instanceof RubyString)) {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        return false;
+    }
+
+
+    private void setTagsAndFailTags(final Object value) {
+        tag(TAGS_FAILURE_TAG);
+        tagFailTags(Valuefier.convert(value));
+    }
+
+    /**
+     * handle key/val pair in `tags`, add _tagsparsefailure to `tags`.
+     * rename FieldReference from `tags` to `_tags`. If existing `_tags` is a list, append value to `_tags`,
+     * eg. [tags][key] -> [_tags][list_index][key]
+     * @param field FieldReference of `tags` with map path
+     * @param value
+     */
+    private void setTagsAndFailTagsWithMap(final FieldReference field, final Object value) {
+        tag(TAGS_FAILURE_TAG);
+
+        List<String> paths = new ArrayList<>();
+        paths.add(TAGS_FAILURE);
+
+        // take list size of _tags as index
+        final Object failTags = Accessors.get(data, TAGS_FAILURE_FIELD);
+        int index = (failTags instanceof List)? ((List) failTags).size() : 0;
+        paths.add(Integer.toString(index));
+
+        // rebuild path & key
+        for(int i = 1; i < field.getPath().length; i++) {
+            paths.add(field.getPath()[i]);
+        }
+        if (field.getKey() != null) {
+            paths.add(field.getKey());
+        }
+
+        String renamedFieldRef = paths.stream().collect(Collectors.joining("][", "[", "]"));
+        final FieldReference renamedField = FieldReference.from(renamedFieldRef);
```

Comment:
@yaauie The rebase idea works well when a single root path is the rebase target. 
eg. `newBase.path == [_tags]`
```
newBase = FieldReference.from("_tags")
oldBase.rebaseOnto(newBase)
```
However, when newBase path is more than one, `newBase.path == [_tags, 0]`, it again needs to construct a string `[_tags][0]` to create newBase `newBase = FieldReference.from("[_tags][0]")` 

Can we make `rebaseOnto()` accept newBase.path?
```java
    public FieldReference rebaseOnto(final List<String> basePath) {
        final List<String> modifiedPath = new ArrayList<>();
        modifiedPath.addAll(basePath);

        if (path.length > 1) { 
            modifiedPath.addAll(Arrays.asList(path).subList(1, path.length)); 
        }

        modifiedPath.add(key);

        return FieldReference.fromTokens(modifiedPath);
    }
```

Reply:
My primary concern with using `FieldReference#from` with a manually-constructed string is that if any component of the path contains reserved characters, the result cannot be parsed and an exception will be thrown, which means including arbitrary field names from the input can lead to runtime exceptions.

When `index` is an `int`, `FieldReference#from(String.format("[_tags][%s]", index))` cannot possibly be invalid.

And `FieldReference.from(String)` is cached, so it is super cheap especially in cases where we use the same input multiple times which we are likely to do with the only variant being numeric indices.

***

I started with a `List<String>`, but there are  _weird_ edge-cases of how `@metadata` fields work that made me suggest taking a `FieldReference` to rebase onto:

| canonical | path | key | type | note |
|-----|------|-----|------|----|
| `[deeply][nested][field]` | `["deeply", "nested"]` | `"field"` | `DATA_CHILD` |  normal |
| `[@metadata][deeply][nested][field]` | `["deeply", "nested"]` | `"field"` | `METADATA_CHILD` | metadata is only  type, not present in path |
| `[@metadata]` | `[]` | `"@metadata"` | `METADATA_PARENT` | huh? metadata is both a type and a key

Even though we are only _using_ this new `FieldReference#rebaseOnto` with known base that is not `@metadata`, I didn't want to introduce a new API as a trap that fails weirdly when someone starts using it in new ways.

_HOWEVER_ my suggestion provides no way to rebase onto the data-root of the event, which can be done by providing an empty list with your implementation.

I think something like the following satisfies the `@metadata` weirdness by extracting the complexity of reconstructing the list of tokens, and also gives us the freedom moving forward to use this for other purposes:

~~~ java
    /**
     * @param shiftLevel the number of levels to shift off of this {@code FieldReference}.
     *                   A value of {@code 0} will _move_ the entire nesting onto the new base
     *                   A value of {@code 1} will remove 1 top-level nesting before moving the remainder
     *                   Specifying more levels than available nestings is an error condition.
     * @param newBase a sequence of zero or more field nestings to prepend
     * @return
     */
    public FieldReference rebaseOnto(@Nonnegative int shiftLevel, final List<String> newBase) {
        if (shiftLevel == 0 && newBase.isEmpty()) { return this; }

        List<String> tokens = toTokens();
        if (tokens.size() <= shiftLevel) {
            throw new IndexOutOfBoundsException(String.format("cannot shift %s levels from field reference with %s tokens", shiftLevel, tokens.size()));
        }

        final List<String> modifiedPath = new ArrayList<>(newBase.size() + tokens.size() - shiftLevel);
        modifiedPath.addAll(newBase);
        modifiedPath.addAll(tokens.subList(shiftLevel, tokens.size()));

        return FieldReference.fromTokens(modifiedPath);
    }

    /**
     * @return a list of tokens that can round-trip through {@link FieldReference#fromTokens}
     *         to produce an identical {@code FieldReference}.
     */
    private List<String> toTokens() {
        // min size to avoid resizing: maybe_meta + path + key
        final List<String> tokens = new ArrayList<>(path.length + 2);

        if (type == META_CHILD) { tokens.add(Event.METADATA); }
        tokens.addAll(Arrays.asList(path));
        tokens.add(key);

        return tokens;
    }
~~~

Reply:
`@metadata` is indeed weird. `rebaseOnto()` accepting path gives much better flexibility 👍